File: 120-triangle/120-triangle.java
class Solution {
    public int minimumTotal(List<List<Integer>> triangle) {
        Map<String,Integer> map = new HashMap<>();
        recur(triangle,0,0,map);
        if(triangle.size()==1){
            return triangle.get(0).get(0);
        }
        return map.get("0,0");
    }
    
    public int recur(List<List<Integer>> triangle,int i,int j, Map<String,Integer> map){
        if(j>triangle.get(i).size()-1 || i>triangle.size()-1){
            return 0;
        }
        if(i==triangle.size()-1){
            return triangle.get(i).get(j);
        }
        String temp = i+","+j;
        if(map.containsKey(temp)){
            return map.get(temp);
        }
        int tempRes = Math.min(triangle.get(i).get(j)+recur(triangle,i+1,j,map),triangle.get(i).get(j)+recur(triangle,i+1,j+1,map));
        map.put(temp,tempRes);
        return tempRes;
        
    }
}

File: 120-triangle/120-triangle.py
class Solution:
    def minimumTotal(self, triangle: List[List[int]]) -> int:
        
        
        def recur(row,col):
            if((row,col) in cache):
                return cache[(row,col)]
            
            res = triangle[row][col]
            if(row+1<len(triangle)):
                p1 = res + recur(row+1,col)
                p2 = res + recur(row+1,col+1)
                res = min(p1,p2)
                
            cache[(row,col)] = res
            return res
          
        cache = dict()
        return recur(0,0)